### PR TITLE
Implement joystick input

### PR DESCRIPTION
**Pull request by smoogipoo**
Axes + buttons are implemented.

`IJoystickState` provices:
* `Buttons`: All the pressed buttons. These include axes as buttons, which is what mania/taiko/catch(?)/osu are all going to use.
* `Axes`: The raw values of all axes.
* `AxisDelta()`: To get the delta for one axis.

Currently supports 20 buttons + 10 axes. We can increase this later if needed but it's already a very extreme use-case.

~Hats not implemented (I don't have a controller to test these with anyway).~

Tested on my FP7 EMP controller.

Supports keybindings (see changes in `InputKey`).

### CODE REVIEW

**Comment by peppy** [comment]
- Hats should probably be mapped to axes. Implementation should be simple.
- We need a deadzone for axes by default. Even with multiple calibrations, it's impossible to get 0.00 on most any analog sticks.

**Comment by peppy** [comment]
Axis to button mapping is currently also very wrong. No matter which axis I trigger, it is always in `Axis 1`

**Comment by peppy** [comment]
I think we may need to have all axes calibrate on startup (on first input frame) to find their zero points. Some controllers (XBOne in this case) have -1 as the default "zero" point for triggers.

File: osu.Framework/Input/JoystickState.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+using System.Collections.Generic;
+
+namespace osu.Framework.Input
+{
+    public class JoystickState : IJoystickState
+    {
+        public IEnumerable<JoystickButton> Buttons { get; set; } = new List<JoystickButton>();
+
+        public IReadOnlyList<float> Axes { get; set; } = new List<float>();
```
Comment (peppy) [inline osu.Framework/Input/JoystickState.cs]: this seems to always be empty

File: osu.Framework/Input/Handlers/Joystick/OpenTKJoystickHandler.cs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+            private float[] hatToAxes(JoystickHatState hatState)
+            {
+                float xAxis = 0;
+                float yAxis = 0;
+
+                if (hatState.IsLeft)
+                    xAxis = -1;
+                else if (hatState.IsRight)
+                    xAxis = 1;
+                if (hatState.IsDown)
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Framework/Input/Handlers/Joystick/OpenTKJoystickHandler.cs]: What do you mean?

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Input/Handlers/Joystick/OpenTKJoystickHandler.cs]: isup became positive

[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Framework/Input/Handlers/Joystick/OpenTKJoystickHandler.cs]: What would you expect it to be?

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Input/Handlers/Joystick/OpenTKJoystickHandler.cs]: The same as every other axis button ._.

**Pull request merged**